src/font_renderer/ct_font_renderer.cpp

```diff
@@ -0,0 +1,140 @@
+
+#if DPFB_USE_CORETEXT
+
+#include "font_renderer/font_renderer.h"
+
+#include <CoreFoundation/CoreFoundation.h>
+#include <CoreGraphics/CoreGraphics.h>
+#include <CoreText/CoreText.h>
+
+
+
+class CtFontRenderer : public FontRenderer {
+public:
+    explicit CtFontRenderer(const FontRendererArgs& args);
+
+    FontMetrics getFontMetrics() const override;
+
+    GlyphIndex getGlyphIndex(char32_t cp) const override;
+    GlyphMetrics getGlyphMetrics(GlyphIndex glyphIdx) const override;
+    void renderGlyph(GlyphIndex glyphIdx, Image& image) const override;
+    virtual ~CtFontRenderer() override;
+private:
+    CTFontRef font;
+};
+
+
+CtFontRenderer::CtFontRenderer(const FontRendererArgs& args)
+{
+    CFStringRef fontName = CFStringCreateWithCString(NULL, (const char*)args.data, kCFStringEncodingUTF8);
+    font = CTFontCreateWithName(fontName, args.pxSize, NULL);
+    CFRelease(fontName);
+    if (!font)
+        throw FontRendererError("CoreText can't init font");
+}
+
+
+FontMetrics CtFontRenderer::getFontMetrics() const
+{
+    FontMetrics metrics;
+
+    metrics.ascender = ceil(CTFontGetAscent(font));
+    metrics.descender = ceil(CTFontGetDescent(font));
+    metrics.lineHeight = ceil(CTFontGetLeading(font) + CTFontGetAscent(font) + CTFontGetDescent(font));
+
+    return metrics;
+}
+
+
+GlyphIndex CtFontRenderer::getGlyphIndex(char32_t cp) const
+{
+    CGGlyph glyph;
+    UniChar ch = (UniChar)cp;
+    if (CTFontGetGlyphsForCharacters(font, &ch, &glyph, 1))
+        return (GlyphIndex)glyph;
+    return 0;
+}
+
+#define CORETEXT_SHIFT_X_POSITION 1
+
+GlyphMetrics CtFontRenderer::getGlyphMetrics(GlyphIndex glyphIdx) const
+{
+    GlyphMetrics glyphMetrics;
+    CGGlyph glyph = (CGGlyph)glyphIdx;
+
+    CGSize advance;
+    CTFontGetAdvancesForGlyphs(font, kCTFontOrientationDefault, &glyph, &advance, 1);
+    glyphMetrics.advance = round(advance.width);
+
+    CGRect boundingRect = CTFontGetBoundingRectsForGlyphs(font, kCTFontOrientationDefault, &glyph, NULL, 1);
+#if CORETEXT_SHIFT_X_POSITION
+    glyphMetrics.size.w = ceil(boundingRect.size.width) + ceil(boundingRect.origin.x);
+#else
+    glyphMetrics.size.w = ceil(boundingRect.size.width);
+#endif
+    glyphMetrics.offset.x = ceil(boundingRect.origin.x);
+
+    glyphMetrics.size.h = ceil(boundingRect.size.height);
+    glyphMetrics.offset.y = ceil(boundingRect.size.height) + ceil(boundingRect.origin.y);
```

Comment:
I wonder why boundingRect.origin contributes to size of the glyph's bitmap? Have you tried to render the text with `tools/draw-text.py`?

Reply:
I don't know about FreeType and STB, but CoreText draws and calculates everything in floating-point. Most glyphs have non-integer offsets, and if we shift them back so they are rendered from point 0 in our bitmap (by setting `position.x=-origin.x`), they are displayed slightly different from their original shape. That's why I set `position.x=0`, so the glyph is rendered starting at its `origin.x`. But then it's needed to increase the width, so it wouldn't get cut, and set our local origin to 0 (which I forgot to do and now fixed by 88c07d86b0eac54bfb310c7badff4dc248bd59f3), because the true origin (floating-point) is already included in our bitmap. I've illustrated it with a drawing:
![IMG_5782](https://user-images.githubusercontent.com/446776/81510147-5b640080-9318-11ea-84c9-d91d63ea81aa.jpeg)

Here is an image which was generated by `draw-text.py` with `CORETEXT_SHIFT_X_POSITION=0`:
![mhd_0](https://user-images.githubusercontent.com/446776/81510104-29529e80-9318-11ea-8092-18b66b08fdb8.png)
Here is an image which was generated by `draw-text.py` with `CORETEXT_SHIFT_X_POSITION=1`:
![mhd_1](https://user-images.githubusercontent.com/446776/81510126-34a5ca00-9318-11ea-9205-d1f097d41299.png)
The second one looks almost like rendered by Core Text.

Reply:
Both x and y offsets should be in GlyphMetrics::offset, so that they are added during drawing and not baked right into a glyph bitmap.

Reply:
They are integer, we can't shift a bitmap by less than a pixel. It's done like you want when `CORETEXT_SHIFT_POSITION=0` and you see what happens.